### ISSUE TITLE
DataObjectManager.open promoting mode 'w' to mode 'w+' when opening existing data object

### ISSUE THREAD

**Issue by tedgin** [irods/python-irodsclient#393]
I'm using python-irodsclient version 1.1.4 with iRODS 4.2.10. I use `irods.manager.DataObjectManager.open` to open a data object using open mode `w` and implicitly allow the data object to be created, If the data object already exists, the library appears to change the open mode to `w+`. This changes the access to the data object from write-only to read-write.  

Here's an example. I run the following Python script.

```python
import os
from irods.session import iRODSSession

OBJ = "/tempZone/home/centos/demo-obj"
ENV = os.path.expanduser('~/.irods/irods_environment.json')

with iRODSSession(irods_env_file=ENV) as session:
    if session.data_objects.exists(OBJ):
        session.data_objects.unlink(path=OBJ, force=True)

    # data object is created during this call   
    session.data_objects.open(path=OBJ, mode='w')

    # data object is exists when this is called
    session.data_objects.open(path=OBJ, mode='w')
```

I have the following rule attached `pep_api_data_obj_open_post`.

```
pep_api_data_obj_open_post(*Instance, *Comm, *DataObjInp) {
  writeLine(
    'serverLog',
    'pep_api_data_obj_open_post: DataObjInp.obj_path = "' ++ *DataObjInp.obj_path ++ '"' );
  writeLine(
    'serverLog',
    'pep_api_data_obj_open_post: DataObjInp.open_flags = "' ++ *DataObjInp.open_flags ++ '"' );
}
```

When I run the script the rule writes the following to the rodsLog.

```
Sep 15 23:01:20 pid:5260 NOTICE: writeLine: inString = pep_api_data_obj_open_post: DataObjInp.obj_path = "/tempZone/home/centos/demo-obj"
Sep 15 23:01:20 pid:5260 NOTICE: writeLine: inString = pep_api_data_obj_open_post: DataObjInp.open_flags = "577"
Sep 15 23:01:21 pid:5260 NOTICE: writeLine: inString = pep_api_data_obj_open_post: DataObjInp.obj_path = "/tempZone/home/centos/demo-obj"
Sep 15 23:01:21 pid:5260 NOTICE: writeLine: inString = pep_api_data_obj_open_post: DataObjInp.open_flags = "578"
```

The flag set `577` translates to `O_WRONLY | O_CREAT | O_TRUNC`, which is the definition of open mode `w` with create allowed.  The flag set `578` translates to `O_RDWR | O_CREATE | O_TRUNC`, which is the definition of open mode `w+` with create allowed.

**Comment by tedgin** [comment]
This appears to be true for open mode `a` as well.

**Comment by d-w-moore** [comment]
I wonder if this is a consequence of the linear model of permissions in iRODS.  For the iRODS server in general, writeability implies readability, which of course is the only difference between `a` and `a+` , or as the case may be, `w` and `w+`. @alanking ?

**Comment by alanking** [comment]
I can't really tell if this is a bug or not, but it is clearly intentional in `rsDataObjOpen` in the iRODS server for an overwrite of a replica. I would also add that this has been the case in iRODS servers back to 4.1.12, perhaps even before that.

As you can see here, `'w'` is correctly being translated as `O_WRONLY` in `irods.manager.data_object_manager.open`: https://github.com/irods/python-irodsclient/blob/e393b1104624f6427cddb6487fc0e346b832caa6/irods/manager/data_object_manager.py#L305

However, the `create` parameter - which is unset in @tedgin's PRC code - defaults to `True`. That means we are going to overwrite the existing replica.

When an existing replica is opened with `O_CREAT`, it is interpreted as an overwrite of that replica. The `O_WRONLY` is explicitly bitmasked away and `O_RDWR` is used instead (link is for 4.2.10): https://github.com/irods/irods/blob/f615a350dc0b787e9140a61c501fdd7e765b40dc/server/api/src/rsDataObjOpen.cpp#L199-L209

I think that the issue presented here is no new. `O_RDWR` has been bitmask-or'd into the `openFlags` for overwrites for a long time (link is for 4.1.12): https://github.com/irods/irods/blob/8e71ad58c2c9341d9bc3a1807c8487233bd2a6d0/iRODS/server/api/src/rsDataObjCreate.cpp#L204

`O_WRONLY` started being explicitly bitmasked away in this commit, in service of irods/irods#5546: https://github.com/irods/irods/commit/d4232f043491a0c0d77eff27314f4111acca4b1e

So... if this is not the expected behavior, then I think the fix would be in the server, not here.

**Comment by d-w-moore** [comment]
> I think the fix would be in the server, not here.

Agreed.  I think it is also fair to say this is only a bug if we are promising our permissions model is 100% POSIX conformant. Already with the OWN > WRITE > READ > NULL ordering, I don't think we can claim such.

**Comment by korydraughn** [comment]
After some discussion, the plan is to see if NOT changing the access mode affects anything.

@tedgin Do you feel the current behavior could lead to issues?

**Comment by korydraughn** [comment]
Using https://github.com/korydraughn/irods/commit/cbe48243aab9db9ff050621415afbb6218a34493, I've verified that the C++ unit tests pass without any issues.

I'm now running the core tests to see what happens.

**Comment by tedgin** [comment]
The current behavior may allow an iRODS rule to write to a replica that has been read-locked at the iRODS level.

**Comment by trel** [comment]
A rule always executes as the user… so no different than via an API call… 

Tony are you saying this behavior is a feature / useful today?  Or a bug that is ‘fixed’ by our proposed change?

**Comment by korydraughn** [comment]
Using https://github.com/korydraughn/irods/commit/cbe48243aab9db9ff050621415afbb6218a34493, I've verified that the core tests pass.

Given that the unit tests and core test suite pass, it appears nothing requires that the access mode be changed.

**Comment by korydraughn** [comment]
This issue needs to be moved to irods/irods as it applies to all clients.

**Comment by trel** [comment]
> This issue needs to be moved to irods/irods as it applies to all clients.

Agreed.
